src/progress.rs

```diff
@@ -47,118 +47,155 @@ impl Default for ProgressState {
+        self.progress.get(&id)
     }
 
     /// Grabs a mutable reference to the progress of a node.
+    #[inline]
     pub fn get_mut(&mut self, id: u64) -> Option<&mut Progress> {
-        let progress = self.voters.get_mut(&id);
-        if progress.is_none() {
-            return self.learners.get_mut(&id);
-        }
-        progress
+        self.progress.get_mut(&id)
     }
 
     /// Returns an iterator across all the nodes and their progress.
-    pub fn iter(&self) -> impl Iterator<Item = (&u64, &Progress)> {
-        self.voters.iter().chain(&self.learners)
+    #[inline]
+    pub fn iter(&self) -> impl ExactSizeIterator<Item = (&u64, &Progress)> {
+        self.progress.iter()
     }
 
     /// Returns a mutable iterator across all the nodes and their progress.
-    pub fn iter_mut(&mut self) -> impl Iterator<Item = (&u64, &mut Progress)> {
-        self.voters.iter_mut().chain(&mut self.learners)
+    #[inline]
+    pub fn iter_mut(&mut self) -> impl ExactSizeIterator<Item = (&u64, &mut Progress)> {
+        self.progress.iter_mut()
     }
 
     /// Adds a voter node
-    pub fn insert_voter(&mut self, id: u64, pr: Progress) -> Result<(), Error> {
-        if self.voters.contains_key(&id) {
-            Err(Error::Exists(id, "voters"))?
-        }
-        if self.learners.contains_key(&id) {
+    pub fn insert_voter(&mut self, id: u64, mut pr: Progress) -> Result<(), Error> {
+        if self.learner_ids().contains(&id) {
             Err(Error::Exists(id, "learners"))?;
+        } else if self.voter_ids().contains(&id) {
+            Err(Error::Exists(id, "voters"))?;
         }
-        self.voters.insert(id, pr);
+        pr.is_learner = false;
+        self.configuration.voters.insert(id);
+        self.progress.insert(id, pr);
         Ok(())
     }
 
     /// Adds a learner to the cluster
-    pub fn insert_learner(&mut self, id: u64, pr: Progress) -> Result<(), Error> {
-        if self.voters.contains_key(&id) {
-            Err(Error::Exists(id, "voters"))?
-        }
-        if self.learners.contains_key(&id) {
-            Err(Error::Exists(id, "learners"))?
+    pub fn insert_learner(&mut self, id: u64, mut pr: Progress) -> Result<(), Error> {
+        if self.learner_ids().contains(&id) {
+            Err(Error::Exists(id, "learners"))?;
+        } else if self.voter_ids().contains(&id) {
+            Err(Error::Exists(id, "voters"))?;
         }
-        self.learners.insert(id, pr);
+        pr.is_learner = true;
+        self.configuration.learners.insert(id);
+        self.progress.insert(id, pr);
         Ok(())
     }
 
     /// Removes the peer from the set of voters or learners.
     pub fn remove(&mut self, id: u64) -> Option<Progress> {
-        match self.voters.remove(&id) {
-            None => self.learners.remove(&id),
-            some => some,
-        }
+        self.configuration.voters.remove(&id);
+        self.configuration.learners.remove(&id);
+        self.progress.remove(&id)
```

Comment:
Could you add few asserts in `fn remove` and `fn insert` to make sure that the `configuration` is consistent with the `progress`?